File: app/src/main/java/com/blankj/androidutilcode/feature/core/app/AppActivity.java
package com.blankj.androidutilcode.feature.core.app;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.View;
import android.widget.TextView;

import com.blankj.androidutilcode.Config;
import com.blankj.androidutilcode.R;
import com.blankj.androidutilcode.base.BaseBackActivity;
import com.blankj.androidutilcode.helper.PermissionHelper;
import com.blankj.utilcode.util.AppUtils;
import com.blankj.utilcode.util.FileUtils;
import com.blankj.utilcode.util.LogUtils;
import com.blankj.utilcode.util.SpanUtils;
import com.blankj.utilcode.util.ToastUtils;
import com.blankj.utilcode.util.Utils;

/**
 * <pre>
 *     author: Blankj
 *     blog  : http://blankj.com
 *     time  : 2016/10/13
 *     desc  : demo about AppUtils
 * </pre>
 */

public class AppActivity extends BaseBackActivity {

    private final OnReleasedListener listener = new OnReleasedListener() {
        @Override
        public void onReleased() {
            AppUtils.installApp(Config.TEST_APK_PATH);
        }
    };

    public static void start(Context context) {
        Intent starter = new Intent(context, AppActivity.class);
        context.startActivity(starter);
    }

    @Override
    public void initData(@Nullable Bundle bundle) {
        AppUtils.registerAppStatusChangedListener(this, new Utils.OnAppStatusChangedListener() {
            @Override
            public void onForeground() {
                ToastUtils.showShort("foreground");
            }

            @Override
            public void onBackground() {
                ToastUtils.showShort("background");
            }
        });
    }

    @Override
    public int bindLayout() {
        return R.layout.activity_app;
    }

    @Override
    public void initView(Bundle savedInstanceState, View contentView) {
        getToolBar().setTitle(getString(R.string.demo_app));

        findViewById(R.id.btn_install_app).setOnClickListener(this);
        findViewById(R.id.btn_install_app_silent).setOnClickListener(this);
        findViewById(R.id.btn_uninstall_app).setOnClickListener(this);
        findViewById(R.id.btn_uninstall_app_silent).setOnClickListener(this);
        findViewById(R.id.btn_launch_app).setOnClickListener(this);
        findViewById(R.id.btn_relaunch_app).setOnClickListener(this);
        findViewById(R.id.btn_exit_app).setOnClickListener(this);
        findViewById(R.id.btn_launch_app_details_settings).setOnClickListener(this);
        TextView tvAboutApp = findViewById(R.id.tv_about_app);
        tvAboutApp.setText(new SpanUtils()
                .appendLine("isAppRoot: " + AppUtils.isAppRoot())
                .appendLine("isAppDebug: " + AppUtils.isAppDebug())
                .appendLine("isAppSystem: " + AppUtils.isAppSystem())
                .appendLine("isAppForeground: " + AppUtils.isAppForeground())
                .append("getAppIcon: ").appendImage(AppUtils.getAppIcon(), SpanUtils.ALIGN_CENTER)
                .appendLine()
                .appendLine("getAppPackageName: " + AppUtils.getAppPackageName())
                .appendLine("getAppName: " + AppUtils.getAppName())
                .appendLine("getAppPath: " + AppUtils.getAppPath())
                .appendLine("getAppVersionName: " + AppUtils.getAppVersionName())
                .appendLine("getAppVersionCode: " + AppUtils.getAppVersionCode())
                .append("getAppSignatureSHA1: " + AppUtils.getAppSignatureSHA1())
                .create());
    }

    @Override
    public void doBusiness() {

    }

    @Override
    public void onWidgetClick(View view) {
        switch (view.getId()) {
            case R.id.btn_install_app:
                if (AppUtils.isAppInstalled(Config.TEST_PKG)) {
                    ToastUtils.showShort(R.string.app_install_tips);
                } else {
                    PermissionHelper.requestStorage(new PermissionHelper.OnPermissionGrantedListener() {
                        @Override
                        public void onPermissionGranted() {
                            if (!FileUtils.isFileExists(Config.TEST_APK_PATH)) {
                                new ReleaseInstallApkTask(listener).execute();
                            } else {
                                listener.onReleased();
                                LogUtils.d("test apk existed.");
                            }
                        }
                    });
                }
                break;
            case R.id.btn_install_app_silent:
                if (AppUtils.isAppInstalled(Config.TEST_PKG)) {
                    ToastUtils.showShort(R.string.app_install_tips);
                } else {
                    if (AppUtils.installAppSilent(Config.TEST_APK_PATH)) {
                        ToastUtils.showShort(R.string.install_successfully);
                    } else {
                        ToastUtils.showShort(R.string.install_unsuccessfully);
                    }
                }
                break;
            case R.id.btn_uninstall_app:
                if (AppUtils.isAppInstalled(Config.TEST_PKG)) {
                    AppUtils.uninstallApp(Config.TEST_PKG);
                } else {
                    ToastUtils.showShort(R.string.app_uninstall_tips);
                }
                break;
            case R.id.btn_uninstall_app_silent:
                if (AppUtils.isAppInstalled(Config.TEST_PKG)) {
                    if (AppUtils.uninstallAppSilent(Config.TEST_PKG, false)) {
                        ToastUtils.showShort(R.string.uninstall_successfully);
                    } else {
                        ToastUtils.showShort(R.string.uninstall_unsuccessfully);
                    }
                } else {
                    ToastUtils.showShort(R.string.app_uninstall_tips);
                }
                break;
            case R.id.btn_launch_app:
                AppUtils.launchApp(this.getPackageName());
                break;
            case R.id.btn_relaunch_app:
                AppUtils.relaunchApp();
                break;
            case R.id.btn_launch_app_details_settings:
                AppUtils.launchAppDetailsSettings();
                break;
            case R.id.btn_exit_app:
                AppUtils.exitApp();
                break;
        }
    }

    @Override
    protected void onDestroy() {
        AppUtils.unregisterAppStatusChangedListener(this);
        super.onDestroy();
    }
}

File: feature/utilcode/pkg/src/main/java/com/blankj/utilcode/pkg/feature/app/AppActivity.kt
package com.blankj.utilcode.pkg.feature.app

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.blankj.common.activity.CommonActivity
import com.blankj.common.helper.PermissionHelper
import com.blankj.common.item.*
import com.blankj.utilcode.constant.PermissionConstants
import com.blankj.utilcode.pkg.Config
import com.blankj.utilcode.pkg.R
import com.blankj.utilcode.util.*

/**
 * ```
 * author: Blankj
 * blog  : http://blankj.com
 * time  : 2016/10/13
 * desc  : demo about AppUtils
 * ```
 */
class AppActivity : CommonActivity(), Utils.OnAppStatusChangedListener {

    var isRegisterAppStatusChangedListener: Boolean = false

    companion object {
        fun start(context: Context) {
            PermissionHelper.request(context, object : PermissionUtils.SimpleCallback {
                override fun onGranted() {
                    val starter = Intent(context, AppActivity::class.java)
                    context.startActivity(starter)
                }

                override fun onDenied() {
                }
            }, PermissionConstants.STORAGE)
        }
    }

    private val listener = object : OnReleasedListener {
        override fun onReleased() {
            return AppUtils.installApp(Config.TEST_APK_PATH)
        }
    }

    override fun bindTitleRes(): Int {
        return R.string.demo_app
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        LogUtils.e(requestCode, resultCode)
    }

    override fun bindItems(): MutableList<CommonItem<*>> {
        return CollectionUtils.newArrayList(
                CommonItemSwitch("registerAppStatusChangedListener", { isRegisterAppStatusChangedListener }, {
                    isRegisterAppStatusChangedListener = it
                    if (it) {
                        AppUtils.registerAppStatusChangedListener(this)
                    } else {
                        AppUtils.unregisterAppStatusChangedListener(this)
                    }
                }),
                CommonItemTitle("isAppRoot", AppUtils.isAppRoot().toString()),
                CommonItemTitle("isAppDebug", AppUtils.isAppDebug().toString()),
                CommonItemTitle("isAppSystem", AppUtils.isAppSystem().toString()),
                CommonItemTitle("isAppForeground", AppUtils.isAppForeground(AppUtils.getAppPackageName()).toString()),
                CommonItemTitle("isAppRunning", AppUtils.isAppRunning(AppUtils.getAppPackageName()).toString()),
                CommonItemImage("getAppIcon") {
                    it.setImageDrawable(AppUtils.getAppIcon())
                },
                CommonItemTitle("getAppPackageName", AppUtils.getAppPackageName()),
                CommonItemTitle("getAppName", AppUtils.getAppName()),
                CommonItemTitle("getAppPath", AppUtils.getAppPath()),
                CommonItemTitle("getAppVersionName", AppUtils.getAppVersionName()),
                CommonItemTitle("getAppVersionCode", AppUtils.getAppVersionCode().toString()),
                CommonItemTitle("getAppSignaturesSHA1", AppUtils.getAppSignaturesSHA1().toString()),
                CommonItemTitle("getAppSignaturesSHA256", AppUtils.getAppSignaturesSHA256().toString()),
                CommonItemTitle("getAppSignaturesMD5", AppUtils.getAppSignaturesMD5().toString()),
                CommonItemTitle("getAppUid", AppUtils.getAppUid().toString()),
                CommonItemTitle("getApkInfo", AppUtils.getApkInfo(AppUtils.getAppPath()).toString()),

                CommonItemClick(R.string.app_install) {
                    if (AppUtils.isAppInstalled(Config.TEST_PKG)) {
                        ToastUtils.showShort(R.string.app_install_tips)
                    } else {
                        if (!FileUtils.isFileExists(Config.TEST_APK_PATH)) {
                            ReleaseInstallApkTask(listener).execute()
                        } else {
                            listener.onReleased()
                        }
                    }
                },
                CommonItemClick(R.string.app_uninstall) {
                    if (AppUtils.isAppInstalled(Config.TEST_PKG)) {
                        AppUtils.uninstallApp(Config.TEST_PKG)
                    } else {
                        ToastUtils.showShort(R.string.app_uninstall_tips)
                    }
                },
                CommonItemClick(R.string.app_launch) {
                    AppUtils.launchApp(this.packageName)
                },
                CommonItemClick(R.string.app_relaunch) {
                    AppUtils.relaunchApp()
                },
                CommonItemClick(R.string.app_launch_details_settings, true) {
                    AppUtils.launchAppDetailsSettings()
                },
                CommonItemClick(R.string.app_exit) {
                    AppUtils.exitApp()
                }
        )
    }

    override fun onForeground(activity: Activity) {
        ToastUtils.showShort("onForeground\n${activity.javaClass.simpleName}")
    }

    override fun onBackground(activity: Activity) {
        ToastUtils.showShort("onBackground\n${activity.javaClass.simpleName}")
    }

    override fun onDestroy() {
        super.onDestroy()
        if (isRegisterAppStatusChangedListener) {
            AppUtils.unregisterAppStatusChangedListener(this)
        }
    }
}

class ReleaseInstallApkTask(private val mListener: OnReleasedListener) : ThreadUtils.SimpleTask<Unit>() {

    override fun doInBackground() {
        ResourceUtils.copyFileFromAssets("test_install", Config.TEST_APK_PATH)
    }

    override fun onSuccess(result: Unit) {
        mListener.onReleased()
    }

    fun execute() {
        ThreadUtils.executeByIo(this)
    }
}

interface OnReleasedListener {
    fun onReleased()
}